pkg/apis/extensions/types.go

```diff
@@ -82,43 +81,45 @@ type SubresourceReference struct {
+	UtilizationTarget Utilization `json:"utilizationTarget"`
 }
 
 // HorizontalPodAutoscalerSpec is the specification of a horizontal pod autoscaler.
 type HorizontalPodAutoscalerSpec struct {
 	// ScaleRef is a reference to Scale subresource. HorizontalPodAutoscaler will learn the current resource consumption from its status,
 	// and will set the desired number of pods by modyfying its spec.
-	ScaleRef *SubresourceReference `json:"scaleRef"`
+	ScaleRef SubresourceReference `json:"scaleRef"`
 	// MinReplicas is the lower limit for the number of pods that can be set by the autoscaler.
-	MinReplicas int `json:"minReplicas"`
+	MinReplicas *int `json:"minReplicas"`
 	// MaxReplicas is the upper limit for the number of pods that can be set by the autoscaler. It cannot be smaller than MinReplicas.
 	MaxReplicas int `json:"maxReplicas"`
-	// Target is the target average consumption of the given resource that the autoscaler will try to maintain by adjusting the desired number of pods.
-	// Currently two types of resources are supported: "cpu" and "memory".
-	Target ResourceConsumption `json:"target"`
+	// CPU utilization is the target average CPU utilization (represented as a fraction of requested CPU) over all the pods.
+	CPUUtilization *CPUTargetUtilization `json:"cpuUtilization,omitempty"`
 }
 
 // HorizontalPodAutoscalerStatus contains the current status of a horizontal pod autoscaler
 type HorizontalPodAutoscalerStatus struct {
-	// TODO: Consider if it is needed.
+	// ObservedGeneration is the most recent generation observed by this autoscaler.
+	ObservedGeneration *int64 `json:"observedGeneration,omitempty"`
+
+	// LastScaleTime is the last time the HorizontalPodAutoscaler scaled the number of pods.
+	// This is used by the autoscaler to control how often the number of pods is changed.
+	LastScaleTime *unversioned.Time `json:"lastScaleTime,omitempty"`
+
 	// CurrentReplicas is the number of replicas of pods managed by this autoscaler.
 	CurrentReplicas int `json:"currentReplicas"`
 
 	// DesiredReplicas is the desired number of replicas of pods managed by this autoscaler.
 	DesiredReplicas int `json:"desiredReplicas"`
 
-	// CurrentConsumption is the current average consumption of the given resource that the autoscaler will
-	// try to maintain by adjusting the desired number of pods.
-	// Two types of resources are supported: "cpu" and "memory".
-	CurrentConsumption *ResourceConsumption `json:"currentConsumption"`
-
-	// LastScaleTimestamp is the last time the HorizontalPodAutoscaler scaled the number of pods.
-	// This is used by the autoscaler to controll how often the number of pods is changed.
-	LastScaleTimestamp *unversioned.Time `json:"lastScaleTimestamp,omitempty"`
+	// Current average CPU utilization over all pods, represented as a fraction of requtested CPU,
+	// e.g. 0.7 means that and average pod is using now 70% of its requested CPU.
+	CurrentCPUUtilization *Utilization `json:"currentCPUUtilization"`
```

Comment:
I think Current is redundant here since we don't label the spec value as Desired